### PR TITLE
Refine cancel for read thread stream

### PR DESCRIPTION
**Pull request by SeaRise**
### What problem does this PR solve?

Issue Number: ref #8505

Problem Summary:

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/Storages/DeltaMerge/ReadThread/UnorderedInputStream.h [hunk 1]
```diff
@@ -146,5 +154,7 @@ class UnorderedInputStream : public IProfilingBlockInputStream
     // runtime filter
     std::vector<RuntimeFilterPtr> runtime_filter_list;
     int max_wait_time_ms;
+
+    std::atomic_bool is_stoped = false;
```
Comment (gengliqi) [inline dbms/src/Storages/DeltaMerge/ReadThread/UnorderedInputStream.h:158]:
```suggestion
    std::atomic_bool is_stopped = false;
```

**Review by gengliqi — APPROVED** [review]

File: dbms/src/Storages/DeltaMerge/ReadThread/UnorderedInputStream.h [hunk 2]
```diff
@@ -67,6 +65,16 @@ class UnorderedInputStream : public IProfilingBlockInputStream
     }
 
 protected:
+    void decreaseRefCount()
+    {
+        bool ori = false;
+        if (is_stoped.compare_exchange_strong(ori, true))
```
Comment (SeaRise) [inline dbms/src/Storages/DeltaMerge/ReadThread/UnorderedInputStream.h:71]:
```suggestion
        if (is_stopped.compare_exchange_strong(ori, true))
```

File: dbms/src/Storages/DeltaMerge/ReadThread/UnorderedInputStream.h [hunk 3]
```diff
@@ -46,11 +46,9 @@ class UnorderedInputStream : public IProfilingBlockInputStream
         LOG_DEBUG(log, "Created, pool_id={} ref_no={}", task_pool->pool_id, ref_no);
     }
 
-    ~UnorderedInputStream() override
-    {
-        task_pool->decreaseUnorderedInputStreamRefCount();
-        LOG_DEBUG(log, "Destroy, pool_id={} ref_no={}", task_pool->pool_id, ref_no);
-    }
+    void cancel(bool /*kill*/) override { decreaseRefCount(); }
```
[the remark this answers is not in the document]
Comment (SeaRise) [reply dbms/src/Storages/DeltaMerge/ReadThread/UnorderedInputStream.h:49]: called here
https://github.com/pingcap/tiflash/blob/5417d84de73ba05288de672b58ab60c85c89bd25/dbms/src/Flash/Mpp/MPPTask.cpp#L183-L190

[the remark this answers is not in the document]
Comment (SeaRise) [reply dbms/src/Storages/DeltaMerge/ReadThread/UnorderedInputStream.h:49]: https://github.com/pingcap/tiflash/blob/5417d84de73ba05288de672b58ab60c85c89bd25/dbms/src/Flash/Executor/DataStreamExecutor.cpp#L61-L65

[the remark this answers is not in the document]
Comment (gengliqi) [reply dbms/src/Storages/DeltaMerge/ReadThread/UnorderedInputStream.h:49]: https://github.com/pingcap/tiflash/blob/5417d84de73ba05288de672b58ab60c85c89bd25/dbms/src/DataStreams/MultiplexInputStream.h#L157-L173
It seems that `cancel` will not be called for `UnorderedInputStream` in `shared_pool`.

**Pull request merged**